### DISCUSSION TITLE
false negative?: unpacking Iterable to fixed length positional arguments

### DISCUSSION THREAD

**Discussion by pinterior** [microsoft/pyright#8618] [Q&A] [answered]
current Pyright (1.1.374) does not generate errors for unpacking Iterable to fixed length arguments.
i think it could be helpful if errors reported.

#### simple case

```python
from collections.abc import Iterable
def f(): pass
def g(x: Iterable[str]): f(*x)
```
[pyright-play.net](https://pyright-play.net/?strict=true&code=MQBwngTglg5gFgFwFwAIDODoGMECgBmEA9gLYpZEA2lApjlEQHZoB0AhgEZYpQkhEQEKAJIIaETrVwATGvhT4AFAEpUINmjQy5KGIoAeqUeMk0A2hggBdVQsUAqfctxA)

#### common typo case

function forwarding kwargs, but mistyped `**` to `*`.

```python
from typing import TypedDict, Unpack

class A(TypedDict, total=False):
    x: int

def f(**kwargs: Unpack[A]):
    pass

def g(**kwargs: Unpack[A]):
    f(*kwargs)  # common typo: single asterisk

g(x=3)  # TypeError: f() takes 0 positional arguments but 1 was given
```
[pyright-play.net](https://pyright-play.net/?strict=true&code=MQAgDgngTglg5gCwC4C4QGcmwMZIFB4BmUA9gLYhIRgwB2cIMZYJUSIAKtQKYAmAIjFwAaEAFVaYAIbYA1gWwAbKenQgAggAouYPoJGUSSKYoC8AMRPpuAShR4QjkAA80dfHl7dCIQpoBU-rIA7lJQcOhoEtJyANrqALp2Dk7SqgRePnABQaHhkeKSMrLxSfZOvgEhYRE2BNnOpgDMdUA)

**Comment by erictraut** [comment, marked as answer]
This behavior is intended, so I don't consider this a bug. The python typing spec does not specify whether a type checker should generate an error when using an unpack operator on an operand of unknown length. This is a behavior that should arguably be specified one way or another in the spec, but currently it is not. Pyright's current behavior is based on two principles:
1. When code is potentially correct but also potentially a bug, pyright opts to avoid a false positive over a false negative. In this case, there's a false negative, but reporting this condition would result in false positives in valid Python code.
2. Unless there's good reason to deviate from behaviors established by mypy, pyright sticks to mypy's behaviors. Mypy doesn't generate an error, at least in the first "simple case" example. It does generate an error in your second example, but that means mypy is internally inconsistent. I'm not sure what principle it is using here.

For these reasons, I don't plan to modify the behavior here — at least not now. If the typing spec is updated or there is some consensus in the community that this condition should generate a type error, I'm open to changing this behavior. If you'd like to start a discussion about this, I recommend posting to the [Python typing forum](https://discuss.python.org/c/typing/32).

**Reply by pinterior** [reply]
I see. Thank you.
